src/core/log_callback.h

```diff
@@ -0,0 +1,24 @@
+#pragma once
+
+#include <string>
+#include <functional>
+
+namespace mavsdk::log {
+
+// defined numeric values can be useful for filtering message by severity
+typedef enum {
+  Debug = 0,
+  Info = 1,
+  Warn = 2,
+  Err = 3
+} Level;
+
+/** @brief User-defined callback for logging. Returning true from this callback
+ * prevents default mavsdk`s logging to stdout. Returning false keeps it.
+ */
+typedef std::function<bool(Level level, const std::string &message,
+                           const std::string &file, int line)> Callback;
+
+extern Callback callback;
```

Comment:
Instead of `extern` I would prefer a mechanism with a clear API in `mavsdk.h`, e.g. `Mavsdk::subscribe_log(Callback)`

Reply:
Given the capitalisation of Mavsdk in you message, I assume you mean static method? Because now logging isn't bound to Mavsdk instances, and probably some logging happens in places where no Mavsdk instances exist.
If my assumption is correct, then it's probably a question of style/convention - should I place callback setting into Mavsdk class as static variable/method, or place it into some namespace. I decided that logging belongs to the whole framework and not to the Mavsdk class or any other class, so placed it into separate namespace mavsdk::log. But I'm new here and I'll be glad to hear from you how it should be done correctly.

Regarding extern var vs. subscribe_log -  does subscribe imply that several callbacks are possible at once? Or subscribe_log should just overwrite old callback, being essentially a setter for single callback variable?

Reply:
Honestly, I'd better left implementation of saving messages to a logfile to the users. I believe In many cases it will suffice to just redirect stdout to a file in run time. When I proposed this change I had in mind slightly different case - when the rest of project already has some logging system and user wants to integrate MAVSDK logs into it. E.g. in the project I'm working on we use spdlog framework, and it has several sinks at once - SQLite db, stdout, WebSocket broadcast.

Reply:
> and probably some logging happens in places where no Mavsdk instances exist.

Interesting point. Are you sure? I can't think of any. Generally `Mavsdk` needs to be instantiated for the whole time.

> Or subscribe_log should just overwrite old callback, being essentially a setter for single callback variable?

Yes, basically a setter. That's how they are all implemented, for now, which is not optimal.

One problem we have with the `subscribe_log` API is that MAVSDK prints its version right on loading, so we would have to add the log callback to the constructor. Would that work?

> Honestly, I'd better left implementation of saving messages to a logfile to the users. I believe In many cases it will suffice to just redirect stdout to a file in run time.

That's fine, at least for now that's really good anyway.

>

Reply:
> Interesting point. Are you sure? I can't think of any. Generally Mavsdk needs to be instantiated for the whole time.

Actually I'm thinking that one day, a `System` may be a standalone thing that does not rely on `Mavsdk` anymore. That would be such a case, right?

Reply:
> Interesting point. Are you sure? I can't think of any. Generally Mavsdk needs to be instantiated for the whole time.
For example, any static helper function. Or any lambda that doesn't need to capture Mavsdk otherwise. I'm not that familiar with the codebase to say how often are such cases. After few minutes of searching I found several e.g. LogError() is used in functions 
```
static rpc::follow_me::FollowMeResult::Result translateToRpcResult(const mavsdk::FollowMe::Result& result)
```
```
bool CurlWrapper::download_file_to_path
```
Also LogDebug is used in integration tests.

> One problem we have with the subscribe_log API is that MAVSDK prints its version right on loading, so we would have to add the log callback to the constructor. Would that work?
That's what I'm talking about - because logging logically belongs to the framework and to not the Mavsdk instances (which can exist in any quantity including zero), if we bind logging to Mavsdk instances we may run into any kind of weird complications.

Setter function for the callback is good idea anyway, for the same reasons we usually make setters for class members - e.g. one day we may want to do some additional work on callback change, or validate something. I'll add it

Reply:
Hm, you're right! I did think of the integration tests. An given what @JonasVautherin said your approach then makes sense.